### DISCUSSION TITLE
LabelSampler does not yield balanced batches

### DISCUSSION THREAD

**Discussion by malteekj** [TorchIO-project/torchio#457] [Q&A] [answered]
I'm trying to use the LabelSampler for a segmentation task with the patch pipeline. I want to get balanced batches between patches with and without a ROI. Preferable perfectly balanced each batch (50% with, %50 without). 
However, when I use a batch size of 12, I get a mean of 8.7 per batch, instead of 6 as wanted.  I hope you can help to point out where the problem is.

My ROIs are saved as 3D arrays, where 0 is non-tumor and 1 is tumor, i.e. they are not one-hot encoded.

My code is as follows:
```python
def numpy_reader_float(path):
    return torch.from_numpy(np.load(path)).type(torch.FloatTensor), np.eye(4)

def numpy_reader_long(path):
    return torch.from_numpy(np.load(path)).type(torch.LongTensor), np.eye(4)

# Train set
train_subjects = []
for path in train_paths:
    CT_name = [name for name in os.listdir(path) if 'CT' in name][0]
    CT_path = os.path.join(path, CT_name)
    
    ROI_name = [name for name in os.listdir(path) if 'ROIs' in name][0]
    ROI_path = os.path.join(path, ROI_name)
    
    subj = tio.Subject(
        CT = tio.ScalarImage(CT_path, reader=numpy_reader_float),
        label = tio.LabelMap(ROI_path, reader=numpy_reader_long),
    )
    train_subjects.append(subj)

# Test set
test_subjects = []
for path in test_paths:
    CT_name = [name for name in os.listdir(path) if 'CT' in name][0]
    CT_path = os.path.join(path, CT_name)
    
    ROI_name = [name for name in os.listdir(path) if 'ROIs' in name][0]
    ROI_path = os.path.join(path, ROI_name)
    
    subj = tio.Subject(
        CT = tio.ScalarImage(CT_path, reader=numpy_reader_float),
        label = tio.LabelMap(ROI_path, reader=numpy_reader_long),
    )
    test_subjects.append(subj)
    
'''
Transform
'''
training_transform = tio.Compose([
    tio.RescaleIntensity((0, 1)),
    ])

test_transform = tio.Compose([
    tio.RescaleIntensity((0, 1)),
    ])

'''
Create SubjectDataset
'''
CT_dataset_train = tio.SubjectsDataset(train_subjects, transform=training_transform)
CT_dataset_test = tio.SubjectsDataset(test_subjects, transform=test_transform)

'''
Make the sampler
'''
training_batch_size = 12
test_batch_size = 2 * training_batch_size

patch_size = 64
samples_per_volume = 20
max_queue_length = 300
sampler = tio.data.LabelSampler(patch_size=patch_size, label_probabilities={0: 0.5, 1: 0.5}, label_name='label')

'''
Make Queue (the Dataset wrapper)
'''
patches_training_set = tio.Queue(
    subjects_dataset=CT_dataset_train,
    max_length=max_queue_length,
    samples_per_volume=samples_per_volume,
    sampler=sampler,
    num_workers=8, 
    shuffle_subjects=True,
    shuffle_patches=True,
)

patches_test_set = tio.Queue(
    subjects_dataset=CT_dataset_test,
    max_length=max_queue_length,
    samples_per_volume=samples_per_volume,
    sampler=sampler,
    num_workers=4, 
    shuffle_subjects=True,
    shuffle_patches=True,
)


'''
Wrap in dataloader
'''
CT_dataloader_train = torch.utils.data.DataLoader(
    patches_training_set, batch_size=training_batch_size)

CT_dataloader_test = torch.utils.data.DataLoader(
    patches_test_set, batch_size=test_batch_size)

```

**Comment by fepegar** [comment, marked as answer]
Hi, @malteekj. Your code looks fine. You don't need to use `Compose` for a single transform.

How are you computing that 8.7? It's the mean of what, specifically? And from how many patches? The idea is that *approximately* half of your patches will contain a labeled voxel in the middle and half will not.

**Reply by fepegar** [reply]
I've taken a closer look at the samplers and they are indeed probably biased towards non-background classes, as probabilities for voxels near the boundaries are zeroed in the `WeightedSampler`. That would explain that you get a mean > 0.5 (although 8.7 seems very high). In this example, I get around 0.52:

```python
import torch
import torchio as tio

size = 16
tensor = torch.zeros(1, size, size, 1)
tensor[0, 4, 5:8, 0] = 1
label = tio.LabelMap(tensor=tensor)

subject = tio.Subject(label=label)
patch_size = 2, 2, 1
sampler = tio.LabelSampler(patch_size, label_probabilities={0: 1, 1:1})
values = torch.as_tensor([patch.label.data[0, 1, 1, 0] for patch in sampler(subject, 1000)])
print(values.mean())
```

This is a small bug, I'll create an issue about it.

**Reply by fepegar** [reply]
More info in [this gist](https://gist.github.com/fepegar/9e029c85827f48360ecc1b62b2530fa7)

**Reply by malteekj** [reply]
Thank for the answer @fepegar !
Yes, 8.7 is the mean number of patches containing a ROI when using a batch size of 12 (average of 100 batches).

Can I do anything to get more balanced batches or can I provide more code or sample images that might indicate why this is happening? It really adversely affects the training of the my network, so any help would be greatly appreciated.

**Reply by fepegar** [reply]
Sorry, I forgot to mention that I've worked on this (https://github.com/fepegar/torchio/issues/458) and you should get more balanced patches in `v0.18.28`.

**Reply by fepegar** [reply]
Just in case: you should be counting the patches with the label in the middle voxel (`np.array(patch.shape) // 2)`), not in any position. I think that's why you were getting such a high number.

**Reply by malteekj** [reply]
I tried with counting the center of the patch instead and it did help, but the distribution is still skewed towards more ROIs than no ROIs. The mean was 7.27 from 1000 batches (batch size 12).
Here is a histogram of the distribution:
![image](https://user-images.githubusercontent.com/44469884/107781619-5f5dfc80-6d48-11eb-90d0-65c21f07fd37.png)

**Reply by fepegar** [reply]
If you can provide a minimal working example to reproduce this, I'll be happy to investigate further.

**Reply by fepegar** [reply]
Is this after upgrading to the newest TorchIO version?

**Reply by malteekj** [reply]
Sorry, I have updated to v0.18.28 now, and it is fixed!
mean is 6.008 (batch size 12)

Histogram:
![image](https://user-images.githubusercontent.com/44469884/107785250-a51cc400-6d4c-11eb-9dc5-a98c016709c5.png)

Thanks for the help!

**Reply by fepegar** [reply]
No worries! Thank you for reporting.
